app/api/authentication.py

```diff
@@ -0,0 +1,122 @@
+# coding=utf-8
+from time import localtime, strftime
+
+from flask import request, jsonify, g
+from flask_httpauth import HTTPBasicAuth
+
+from . import api, errors
+from .custom_exceptions import ServerException, ValidationException
+from.decorators import catch_exceptions
+from .utilities import Utilities
+from app.models import User, AnonymousUser, Token
+
+auth = HTTPBasicAuth()
+
+
+@auth.verify_password
+def verify_password(username_or_token, password):
+    """
+    this function verifies a username, password  or token, and determines
+    if a user is an authenticated user or an anonymous user
+    """
+    if username_or_token == "":
+        g.current_user = AnonymousUser()
+        g.token_used = False
+        return True
+    if password == "":
+        g.current_user = User.verify_auth_token(username_or_token)
+        g.token_used = True
+        g.token = username_or_token
```

Comment:
My initial intention was to disallow a user from having multiple valid tokens, but based on the insight you gave below about statelessness, i now see why the initial intention was wrong. Fixed the bug. Thanks for the insight.